core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/TransportCreateSnapshotAction.java

```diff
@@ -87,16 +85,16 @@ public void onResponse() {
                 if (request.waitForCompletion()) {
                     snapshotsService.addListener(new SnapshotsService.SnapshotCompletionListener() {
                         @Override
-                        public void onSnapshotCompletion(SnapshotId snapshotId, SnapshotInfo snapshotInfo) {
-                            if (snapshot.equals(snapshotId.getSnapshot())) {
+                        public void onSnapshotCompletion(Snapshot snapshot, SnapshotInfo snapshotInfo) {
+                            if (snapshot.isSame(request.repository(), request.snapshot())) {
```

Comment:
The listeners in SnapshotService and RestoreService could just themselves implement this matching logic. The API would then be 

```
snapshotsService.addListener(repository, snapshotName, new SnapshotsService.SnapshotCompletionListener() { ... });
```